packages/lib-auth/src/client.js

```diff
@@ -0,0 +1,320 @@
+      refreshToken: tokenData['refresh_token'],
+      scope: tokenData.scope
+    }
+  }
+
+  /**
+   * Checks whether the current access token is still valid or has expired.
+   *
+   * @private
+   * @returns {boolean}
+   */
+  _isAccessTokenValid () {
+    const { accessToken, accessTokenExpiresAt } = this._state
+    return !!accessToken && Date.now() > accessTokenExpiresAt
+  }
+
+  /**
+   * Check whether a user is signed in
+   *
+   * @returns {boolean}
+   */
+  isSignedIn () {
+    return !!this._state.accessToken
+  }
+
+  /**
+   * Makes a request to exchange the refresh token for new token data.
+   *
+   * @async
+   * @private
+   * @returns {Promise} Resolves to the new access token.
+   */
+  async _refreshTokenData () {
+    const data = {
+      'client_id': this._config.clientAppID,
+      'grant_type': 'refresh_token',
+      'refresh_token': this._state.refreshToken
+    }
+    const response = await this._httpClient.post('/oauth/token', data)
+    this._handleNewTokenData(response.data)
+    return this._state.accessToken
+  }
+
+  /**
+   * Registers a new user and logs them in.
+   *
+   * @async
+   * @param {Object} user - The details of the new user account.
+   * @param {boolean} user.betaEmailCommunication - Whether the new user wants to opt in to beta emails.
+   * @param {string} user.creditedName - The new user's credited name (used in citations etc).
+   * @param {string} user.email - The new user's email address.
+   * @param {string} user.login - The new user's username.
+   * @param {boolean} user.globalEmailCommunication - Whether the new user wants to opt in to global emails.
+   * @param {string} user.password - The new user's password.
+   * @param {boolean} user.projectEmailCommunication - Whether the new user wants to opt in to project-related emails when registering via a specific project.
+   * @param {string} user.projectId - The project ID when registering via a specific project.
+   * @returns {Promise} Resolves to a new access token.
+   */
+  async register (user) {
+    if (this.isSignedIn()) {
+      await this.signOut()
+    }
+
+    const csrfToken = await this._getCSRFToken()
+    const data = {
+      'authenticity_token': csrfToken,
+      user: {
+        'beta_email_communication': user.betaEmailCommunication,
+        'credited_name': user.creditedName,
+        email: user.email,
+        'global_email_communication': user.globalEmailCommunication,
+        login: user.login,
+        password: user.password,
+        'project_email_communication': user.projectEmailCommunication,
+        'project_id': user.projectId
+      }
+    }
+    const registerResponse = await this._httpClient.post('/users', data)
+
+    // After the user is created, we need to sign in to actually get an access
+    // token. `panoptes-javascript-client` manages to do it without this step;
+    // there's no `set-cookie` header on the response that I can see, so I'm not
+    // actually sure how.
```

Comment:
My guess it that it doesn't get a token until the first API request is made because every API request is preceded by this check for a token: https://github.com/zooniverse/panoptes-javascript-client/blob/085a94a9b0f028a47db68fff013541b467707a8d/lib/api-client.js#L10-L13